certdb/certdb.go

```diff
@@ -0,0 +1,227 @@
+	RevokedAt time.Time `sql:"revoked_at"`
+	PEM       string    `sql:"pem"`
+}
+
+// OCSPRecord encodes a OCSP response body and its metadata
+// that will be recorded in a database.
+type OCSPRecord struct {
+	Serial string    `sql:"serial"`
+	Body   string    `sql:"body"`
+	Expiry time.Time `sql:"expiry"`
+}
+
+const (
+	insertSQL = `
+INSERT INTO certificates (serial, ca_label, status, reason, expiry, revoked_at, pem)
+	VALUES ($1, $2, $3, $4, $5, $6, $7);`
+
+	selectSQL = `
+SELECT %s FROM certificates
+	WHERE (serial = $1);`
+
+	selectAllSQL = `
+SELECT %s FROM certificates;`
+
+	selectAllUnexpiredSQL = `
+SELECT %s FROM certificates
+WHERE CURRENT_TIMESTAMP < expiry;`
+
+	updateRevokeSQL = `
+UPDATE certificates
+	SET status='revoked', revoked_at=CURRENT_TIMESTAMP, reason=$1
+	WHERE (serial = $2);`
+
+	insertOCSPSQL = `
+INSERT INTO ocsp_responses (serial, body, expiry)
+    VALUES ($1, $2, $3);`
+
+	updateOCSPSQL = `
+UPDATE ocsp_responses
+    SET expiry=$3, body=$2
+	WHERE (serial = $1);`
+
+	selectAllUnexpiredOCSPSQL = `
+SELECT %s FROM ocsp_responses
+WHERE CURRENT_TIMESTAMP < expiry;`
+
+	selectOCSPSQL = `
+SELECT %s FROM ocsp_responses
+    WHERE (serial = $1);`
+)
+
+func wrapCertStoreError(err error) error {
+	if err != nil {
+		return cferr.Wrap(cferr.CertStoreError, cferr.Unknown, err)
+	}
+	return nil
+}
+
+// InsertCertificate puts a CertificateRecord into db.
+func InsertCertificate(db *sql.DB, cr *CertificateRecord) error {
+	res, err := db.Exec(
+		insertSQL,
+		cr.Serial,
+		cr.CALabel,
+		cr.Status,
+		cr.Reason,
+		cr.Expiry.UTC(),
+		cr.RevokedAt,
+		cr.PEM,
+	)
+	if err != nil {
+		return wrapCertStoreError(err)
+	}
+
+	numRowsAffected, _ := res.RowsAffected()
+
+	if numRowsAffected != 1 {
+		return wrapCertStoreError(fmt.Errorf("%d rows are affected, should be 1 row", numRowsAffected))
+	}
+
+	return nil
+}
+
+// GetCertificate gets a CertificateRecord indexed by serial.
+func GetCertificate(db *sql.DB, serial string) (*CertificateRecord, error) {
+	cr := new(CertificateRecord)
+	rows, err := db.Query(fmt.Sprintf(selectSQL, sqlstruct.Columns(*cr)), serial)
+	if err != nil {
+		return nil, wrapCertStoreError(err)
+	}
+	defer rows.Close()
+
+	if rows.Next() {
+		return cr, wrapCertStoreError(sqlstruct.Scan(cr, rows))
+	}
+	return nil, nil
+}
+
+// GetUnexpiredCertificates gets all unexpired certificate from db.
+func GetUnexpiredCertificates(db *sql.DB) (crs []*CertificateRecord, err error) {
+	cr := new(CertificateRecord)
+	rows, err := db.Query(fmt.Sprintf(selectAllUnexpiredSQL, sqlstruct.Columns(*cr)))
+	if err != nil {
+		return nil, wrapCertStoreError(err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		err = sqlstruct.Scan(cr, rows)
+		if err != nil {
+			return nil, wrapCertStoreError(err)
+		}
+		crs = append(crs, cr)
+	}
+
+	return crs, nil
+}
+
+// RevokeCertificate updates a certificate with a given serial number and marks it revoked.
+func RevokeCertificate(db *sql.DB, serial string, reasonCode int) error {
```

Comment:
I am not sure either. The original intent is to abstract sql query with certdb package. I still consider OCSP can be separated from certificate revocation since OCSP is only one of several revocation mechanisms. It left work for OCSP responder/generator which need to look up the certificate table and OCSP table to act correctly.